File: scripts/test/iterm2_panel_switching_test.dart
/// Multi-panel switching test for iTermRemote daemon.
///
/// - Enumerates panels in spatial order (row-major): top-to-bottom, left-to-right
/// - For each panel:
///   - activateSession
///   - captureEvidence (screenshot + crop + red border overlay)
/// - Writes a summary.json into outputDir
///
/// Usage:
///   dart scripts/test/iterm2_panel_switching_test.dart \
///     --ws-url=ws://127.0.0.1:8766 \
///     --output-dir=/tmp/itermremote-panel-switching \
///     --duration=5

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:web_socket_channel/web_socket_channel.dart';

class _WsRpc {
  _WsRpc(this._ws);

  final WebSocketChannel _ws;
  int _seq = 0;
  final Map<String, Completer<Map<String, dynamic>>> _pending = {};

  void start() {
    _ws.stream.listen((message) {
      final obj = jsonDecode(message as String);
      if (obj is! Map<String, dynamic>) return;
      final id = obj['id'];
      if (id is! String) return;
      final c = _pending.remove(id);
      c?.complete(obj);
    });
  }

  Future<Map<String, dynamic>> cmd(
    String target,
    String action,
    Map<String, Object?> payload, {
    Duration timeout = const Duration(seconds: 15),
  }) async {
    final id = '${action}_${_seq++}_${DateTime.now().millisecondsSinceEpoch}';
    final c = Completer<Map<String, dynamic>>();
    _pending[id] = c;

    _ws.sink.add(
      jsonEncode({
        'type': 'cmd',
        'version': 1,
        'id': id,
        'target': target,
        'action': action,
        'payload': payload,
      }),
    );

    return c.future.timeout(
      timeout,
      onTimeout: () {
        _pending.remove(id);
        throw TimeoutException('timeout waiting ack for $target.$action');
      },
    );
  }
}

double _numToDouble(Object? v) {
  if (v is num) return v.toDouble();
  return 0.0;
}

List<Map<String, dynamic>> _sortPanelsSpatial(List<Map<String, dynamic>> panels) {
  // iTerm2 frame is in window coordinates with origin at bottom-left.
  // "Row-major" as user asked: first row then column, which in screen terms means:
  // higher y is visually higher row. So sort y DESC, then x ASC.
  final sorted = [...panels];
  sorted.sort((a, b) {
    final fa = (a['frame'] as Map?)?.cast<String, dynamic>() ?? const {};
    final fb = (b['frame'] as Map?)?.cast<String, dynamic>() ?? const {};
    final ya = _numToDouble(fa['y']);
    final yb = _numToDouble(fb['y']);
    if (ya != yb) return yb.compareTo(ya);
    final xa = _numToDouble(fa['x']);
    final xb = _numToDouble(fb['x']);
    return xa.compareTo(xb);
  });
  return sorted;
}

Future<void> main(List<String> args) async {
  var wsUrl = 'ws://127.0.0.1:8766';
  var outputDir =
      '/tmp/itermremote-panel-switching/${DateTime.now().millisecondsSinceEpoch}';
  var durationSec = 5;

  for (var i = 0; i < args.length; i++) {
    final a = args[i];
    if (a == '--ws-url' && i + 1 < args.length) {
      wsUrl = args[++i];
    } else if (a == '--output-dir' && i + 1 < args.length) {
      outputDir = args[++i];
    } else if (a == '--duration' && i + 1 < args.length) {
      durationSec = int.parse(args[++i]);
    }
  }

  final dir = Directory(outputDir);
  if (!dir.existsSync()) dir.createSync(recursive: true);

  stdout.writeln('[panel_switch] wsUrl=$wsUrl');
  stdout.writeln('[panel_switch] outputDir=$outputDir');
  stdout.writeln('[panel_switch] durationSec=$durationSec');

  final ws = WebSocketChannel.connect(Uri.parse(wsUrl));
  await ws.ready;
  final rpc = _WsRpc(ws);
  rpc.start();

  // list panels
  final listAck = await rpc.cmd('iterm2', 'getSessions', {});
  if (listAck['success'] != true) {
    throw StateError('getSessions failed: ${listAck['error']}');
  }
  final sessionsRaw = (listAck['data'] as Map)['sessions'] as List;
  final panels = sessionsRaw.cast<Map>().map((e) => e.cast<String, dynamic>()).toList();
  final sorted = _sortPanelsSpatial(panels);

  final results = <Map<String, dynamic>>[];

  for (var i = 0; i < sorted.length; i++) {
    final p = sorted[i];
    final title = (p['title'] ?? '').toString();
    final sid = (p['id'] ?? '').toString();
    stdout.writeln('[panel_switch] (${i + 1}/${sorted.length}) activate title=$title sid=$sid');

    final r = <String, dynamic>{
      'order': i + 1,
      'title': title,
      'sessionId': sid,
      'status': 'pending',
      'ts': DateTime.now().millisecondsSinceEpoch,
    };

    try {
      final actAck = await rpc.cmd('iterm2', 'activateSession', {'sessionId': sid});
      if (actAck['success'] != true) {
        throw StateError('activateSession failed: ${actAck['error']}');
      }
      final meta = (actAck['data'] as Map)['meta'] as Map;
      r['cropMeta'] = meta;

      // settle
      await Future<void>.delayed(const Duration(milliseconds: 250));

      // capture
      final capAck = await rpc.cmd(
        'verify',
        'captureEvidence',
        {
          'evidenceDir': outputDir,
          'sessionId': sid,
          'cropMeta': meta,
        },
        timeout: const Duration(seconds: 60),
      );
      if (capAck['success'] != true) {
        throw StateError('captureEvidence failed: ${capAck['error']}');
      }
      final data = capAck['data'] as Map;
      r['screenshotPng'] = data['screenshotPng'];
      r['croppedPng'] = data['croppedPng'];
      r['overlayPng'] = data['overlayPng'];
      r['metaJson'] = data['metaJson'];
      r['status'] = 'success';

      stdout.writeln('[panel_switch]   overlay=${r['overlayPng']}');

      // user asked 5s per panel. We don't yet record video here, but we keep the timing.
      await Future<void>.delayed(Duration(seconds: durationSec));
    } catch (e) {
      r['status'] = 'error';
      r['error'] = e.toString();
      stdout.writeln('[panel_switch]   ERROR: $e');
    }

    results.add(r);
  }

  await ws.sink.close();

  final summary = {
    'ts': DateTime.now().toIso8601String(),
    'wsUrl': wsUrl,
    'order': 'row-major (top-to-bottom, left-to-right) using frame.y desc then frame.x asc',
    'durationSec': durationSec,
    'total': results.length,
    'success': results.where((e) => e['status'] == 'success').length,
    'failed': results.where((e) => e['status'] != 'success').length,
    'results': results,
  };
  File('$outputDir/summary.json')
      .writeAsStringSync(const JsonEncoder.withIndent('  ').convert(summary));

  stdout.writeln('[panel_switch] wrote summary: $outputDir/summary.json');
}

File: scripts/test/iterm2_panel_switching_test.py
#!/usr/bin/env python3
import argparse
import json
import sys
import time
from pathlib import Path

try:
    import websockets
except ImportError:
    print("Missing dependency: websockets. Install via: pip3 install websockets", file=sys.stderr)
    sys.exit(2)


async def ws_cmd(ws, target, action, payload):
    req_id = f"{target}.{action}.{int(time.time()*1000)}"
    msg = {
        "type": "cmd",
        "version": 1,
        "id": req_id,
        "target": target,
        "action": action,
        "payload": payload,
    }
    await ws.send(json.dumps(msg))
    while True:
        raw = await ws.recv()
        data = json.loads(raw)
        if data.get("id") == req_id:
            return data


def sort_panels_spatial(panels):
    # row-major: top-to-bottom, left-to-right
    # Prefer layoutFrame (stable window coords). Fallback to frame if missing.
    # Observed layoutFrame uses y that increases downward, so row order is y ASC.
    def key(p):
        f = p.get("layoutFrame") or p.get("frame") or {}
        y = float(f.get("y", 0))
        x = float(f.get("x", 0))
        # Normalize y to row buckets to reduce tiny jitter
        y_bucket = round(y / 5.0) * 5.0
        return (y_bucket, x)

    return sorted(panels, key=key)


def _load_json(path: Path):
    return json.loads(Path(path).read_text())


def _save_json(path: Path, obj):
    path.write_text(json.dumps(obj, indent=2))


async def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--ws-url", default="ws://127.0.0.1:8766")
    parser.add_argument("--output-dir", default=f"/tmp/itermremote-panel-switching/{int(time.time())}")
    parser.add_argument("--duration", type=int, default=5)
    args = parser.parse_args()

    out_dir = Path(args.output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    results = []

    async with websockets.connect(args.ws_url) as ws:
        list_ack = await ws_cmd(ws, "iterm2", "getSessions", {})
        if not list_ack.get("success"):
            raise RuntimeError(f"getSessions failed: {list_ack}")
        sessions = list_ack["data"]["sessions"]
        panels = sort_panels_spatial(sessions)

        # Use the first panel to fetch cropMeta (cgWindowId + window frames) and capture 1 base screenshot.
        first_sid = panels[0].get("id")
        first_title = panels[0].get("title", "")
        print(f"[0/{len(panels)}] capture base window screenshot via first panel: {first_title} {first_sid}")

        base_act = await ws_cmd(ws, "iterm2", "activateSession", {"sessionId": first_sid})
        if not base_act.get("success"):
            raise RuntimeError(f"activate failed: {base_act}")
        base_meta = base_act["data"]["meta"]
        time.sleep(0.25)

        base_cap = await ws_cmd(
            ws,
            "verify",
            "captureEvidence",
            {
                "evidenceDir": str(out_dir),
                "sessionId": first_sid,
                "cropMeta": base_meta,
            },
        )
        if not base_cap.get("success"):
            raise RuntimeError(f"captureEvidence failed: {base_cap}")

        base_data = base_cap["data"]
        base_screenshot = base_data.get("screenshotPng")
        base_crop_meta = base_meta

        # Write a multi-box overlay image (one screenshot, multiple red boxes + order labels)
        overlay_multi_path = out_dir / "window_multi_overlay.png"
        overlay_multi_json = out_dir / "window_multi_overlay.json"

        overlay_payload = {
            "evidenceDir": str(out_dir),
            "screenshotPng": base_screenshot,
            "windowMeta": {
                "rawWindowFrame": base_crop_meta.get("rawWindowFrame"),
            },
            "panels": [
                {
                    "order": i + 1,
                    "title": p.get("title", ""),
                    "sessionId": p.get("id"),
                    # Use layoutFrame if present (more stable), else frame
                    "frame": p.get("layoutFrame") or p.get("frame"),
                }
                for i, p in enumerate(panels)
            ],
            "outputPng": str(overlay_multi_path),
            "outputJson": str(overlay_multi_json),
        }

        # Ask verify block to render overlay. If not supported, we still proceed per-panel.
        multi_ack = await ws_cmd(ws, "verify", "renderMultiPanelOverlay", overlay_payload)
        if not multi_ack.get("success"):
            print(f"WARN: renderMultiPanelOverlay not available: {multi_ack.get('error')}")
        else:
            print(f"Wrote multi overlay: {overlay_multi_path}")

        # Per-panel switching + evidence
        for idx, p in enumerate(panels, start=1):
            sid = p.get("id")
            title = p.get("title", "")
            print(f"[{idx}/{len(panels)}] activate {title} {sid}")
            entry = {
                "order": idx,
                "title": title,
                "sessionId": sid,
                "status": "pending",
                "ts": int(time.time() * 1000),
            }
            try:
                act = await ws_cmd(ws, "iterm2", "activateSession", {"sessionId": sid})
                if not act.get("success"):
                    raise RuntimeError(f"activate failed: {act}")
                meta = act["data"]["meta"]

                time.sleep(0.25)

                cap = await ws_cmd(
                    ws,
                    "verify",
                    "captureEvidence",
                    {
                        "evidenceDir": str(out_dir),
                        "sessionId": sid,
                        "cropMeta": meta,
                    },
                )
                if not cap.get("success"):
                    raise RuntimeError(f"capture failed: {cap}")

                data = cap["data"]
                entry.update(
                    {
                        "status": "success",
                        "screenshotPng": data.get("screenshotPng"),
                        "croppedPng": data.get("croppedPng"),
                        "overlayPng": data.get("overlayPng"),
                        "metaJson": data.get("metaJson"),
                    }
                )
                print(f"  overlay={entry.get('overlayPng')}")

                time.sleep(args.duration)
            except Exception as e:
                entry["status"] = "error"
                entry["error"] = str(e)
                print(f"  ERROR: {e}")

            results.append(entry)

    summary = {
        "ts": time.strftime("%Y-%m-%dT%H:%M:%S"),
        "wsUrl": args.ws_url,
        "order": "row-major (top-to-bottom, left-to-right) using layoutFrame.y asc then layoutFrame.x asc",
        "durationSec": args.duration,
        "total": len(results),
        "success": len([r for r in results if r["status"] == "success"]),
        "failed": len([r for r in results if r["status"] != "success"]),
        "results": results,
    }

    with open(out_dir / "summary.json", "w") as f:
        json.dump(summary, f, indent=2)

    print(f"Wrote summary: {out_dir / 'summary.json'}")


if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
